modules/e2ee/OlmAdapter.js

```diff
@@ -470,6 +548,175 @@ export class OlmAdapter extends Listenable {
+
+                    return;
+                }
+
+                olmData.sas = new Olm.SAS();
+
+                const { uuid } = msg.data;
+
+                olmData.uuid = uuid;
+
+                const pubKey = olmData.sas.get_pubkey();
+                const olmUtil = new Olm.Utility();
+                const commitment = olmUtil.sha256(pubKey + msg.data);
+
+                olmUtil.free();
+
+                // Send ACCEPT.
+                const ack = {
+                    [JITSI_MEET_MUC_TYPE]: OLM_MESSAGE_TYPE,
+                    olm: {
+                        type: OLM_MESSAGE_TYPES.SAS_ACCEPT,
+                        data: {
+                            uuid,
+                            commitment
+                        }
+                    }
+                };
+
+                this._sendMessage(ack, pId);
+            } else {
+                logger.debug(`Received sas init message from ${pId} but we have no session for them!`);
+
+                this._sendError(participant, 'No session found while processing sas-init');
+            }
+            break;
+        }
+        case OLM_MESSAGE_TYPES.SAS_ACCEPT: {
+            if (olmData.session) {
+                const { commitment, uuid } = msg.data;
+
+                olmData.sasCommitment = commitment;
+
+                const pubKey = olmData.sas.get_pubkey();
+
+                // Send KEY.
+                const ack = {
+                    [JITSI_MEET_MUC_TYPE]: OLM_MESSAGE_TYPE,
+                    olm: {
+                        type: OLM_MESSAGE_TYPES.SAS_KEY,
+                        data: {
+                            key: pubKey,
+                            isInitializer: true,
+                            uuid
+                        }
+                    }
+                };
+
+                this._sendMessage(ack, pId);
+            } else {
+                logger.debug(`Received sas accept message from ${pId} but we have no session for them!`);
+
+                this._sendError(participant, 'No session found while processing sas-accept');
+            }
+            break;
+        }
+        case OLM_MESSAGE_TYPES.SAS_KEY: {
+            if (olmData.session) {
+                if (olmData.sas.is_their_key_set()) {
+                    logger.warn('SAS already has their key!');
+
+                    return;
+                }
+
+                const { key: theirKey, isInitializer, uuid } = msg.data;
+
+                if (olmData.sasCommitment) {
+                    const olmUtil = new Olm.Utility();
+                    const commitment = olmUtil.sha256(theirKey + olmData.startContent);
+
+                    olmUtil.free();
+
+                    if (olmData.sasCommitment !== commitment) {
+                        this._sendError(participant, 'OlmAdapter commitments mismatched');
+                    }
+                }
+
+                olmData.sas.set_their_key(theirKey);
```

Comment:
Can we keep a state variable to not rely on the above check tio stop sending keys indefinitely.